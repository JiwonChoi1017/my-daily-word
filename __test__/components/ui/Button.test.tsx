import {
  Button,
  DoubleButton,
  DuplicateCheckButton,
  ErrorPageButton,
} from "../../../components/ui/Button";
import { render, screen } from "@testing-library/react";

import React from "react";
import userEvent from "@testing-library/user-event";

describe("Buttonコンポーネント", () => {
  const mockOnClick = jest.fn();
  const mockOnClickForSecondButton = jest.fn();

  test("ボタン：文言が想定通りか", () => {
    // Buttonコンポーネントをレンダリング
    render(
      <Button
        className="ButtonClass1"
        text={"Button1"}
        clickHandler={mockOnClick}
      />
    );

    // getByText(): html要素内のテキスト内容を取得
    // toBeInTheDocument(): ドキュメント内に要素が想定通りかを確認
    expect(screen.getByText("Button1")).toBeInTheDocument();
  });

  test("ボタン：cssクラスが想定通りか", async () => {
    // Buttonコンポーネントをレンダリング
    const { container } = render(
      <Button
        className="ButtonClass1"
        text={"Button1"}
        clickHandler={mockOnClick}
      />
    );

    expect(container.firstChild).toHaveClass("ButtonClass1");
  });

  test("ボタン：onClickに渡した関数が呼ばれるか", async () => {
    // Buttonコンポーネントをレンダリング
    render(
      <Button className="button" text={"Button1"} clickHandler={mockOnClick} />
    );

    const button = screen.getByRole("button");
    await userEvent.click(button);

    expect(mockOnClick).toHaveBeenCalled();
  });

  test("ダブルボタン：文言が想定通りか", () => {
    // DoubleButtonコンポーネントをレンダリング
    render(
      <DoubleButton
        button={{
          first: {
            text: "DoubleButton1",
            className: "DoubleButtonClass1",
            clickHandler: mockOnClick,
          },
          second: {
            text: "DoubleButton2",
            className: "DoubleButtonClass2",
            clickHandler: mockOnClickForSecondButton,
          },
        }}
      />
    );

    const doubleButton = screen.getAllByRole("button") as HTMLButtonElement[];

    expect(doubleButton[0].textContent).toEqual("DoubleButton1");
    expect(doubleButton[1].textContent).toEqual("DoubleButton2");
  });

  test("ダブルボタン：cssクラスが想定通りか", () => {
    // DoubleButtonコンポーネントをレンダリング
    render(
      <DoubleButton
        button={{
          first: {
            text: "DoubleButton1",
            className: "DoubleButtonClass1",
            clickHandler: mockOnClick,
          },
          second: {
            text: "DoubleButton2",
            className: "DoubleButtonClass2",
            clickHandler: mockOnClickForSecondButton,
          },
        }}
      />
    );

    const doubleButton = screen.getAllByRole("button") as HTMLButtonElement[];

    expect(doubleButton[0].classList.contains("DoubleButtonClass1")).toBe(true);
    expect(doubleButton[1].classList.contains("DoubleButtonClass2")).toBe(true);
  });

  test("ボタン：onClickに渡した関数が呼ばれるか", async () => {
    // Buttonコンポーネントをレンダリング
    render(
      <Button className="button" text={"Button1"} clickHandler={mockOnClick} />
    );

    const button = screen.getByRole("button");
    await userEvent.click(button);

    expect(mockOnClick).toHaveBeenCalled();
  });

  test("重複チェックボタン：文言が想定通りか", () => {
    // DuplicateCheckButtonコンポーネントをレンダリング
    render(
      <DuplicateCheckButton isDisabled={false} clickHandler={mockOnClick} />
    );

    expect(screen.getByText("重複チェック")).toBeInTheDocument();
  });

  test("重複チェックボタン：cssクラスが想定通りか", async () => {
    // DuplicateCheckButtonコンポーネントをレンダリング
    const { container } = render(
      <DuplicateCheckButton isDisabled={false} clickHandler={mockOnClick} />
    );

    expect(container.firstChild).toHaveClass("duplicateCheckButton");
  });

  test("エラーページのボタン：文言が想定通りか", () => {
    // ErrorPageButtonコンポーネントをレンダリング
    render(
      <ErrorPageButton text={"ErrorPageButton1"} clickHandler={mockOnClick} />
    );

    expect(screen.getByText("ErrorPageButton1")).toBeInTheDocument();
  });

  test("エラーページのボタン：cssクラスが想定通りか", async () => {
    // ErrorPageButtonコンポーネントをレンダリング
    const { container } = render(
      <ErrorPageButton text={"ErrorPageButton1"} clickHandler={mockOnClick} />
    );

    expect(container.firstChild).toHaveClass("button__wrap marginTop20");
  });
});
