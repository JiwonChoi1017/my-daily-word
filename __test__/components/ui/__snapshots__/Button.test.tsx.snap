// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttonのテスト Button: スナップショットテスト 1`] = `
<div>
  <button
    class="ButtonClass1"
    type="button"
  >
    Button1
  </button>
</div>
`;

exports[`Buttonのテスト DoubleButton: スナップショットテスト 1`] = `
<div>
  <div
    class="button__wrap"
  >
    <button
      class="DoubleButtonClass1"
      type="button"
    >
      DoubleButton1
    </button>
    <button
      class="DoubleButtonClass2"
      type="button"
    >
      DoubleButton2
    </button>
  </div>
</div>
`;

exports[`Buttonのテスト DuplicateCheckButton: スナップショットテスト 1`] = `
<div>
  <button
    class="duplicateCheckButton"
    type="button"
  >
    重複チェック
  </button>
</div>
`;

exports[`Buttonのテスト ErrorPageButton: スナップショットテスト 1`] = `
<div>
  <div
    class="button__wrap marginTop20"
  >
    <button
      class="button"
      type="button"
    >
      ErrorPageButton1
    </button>
  </div>
</div>
`;
